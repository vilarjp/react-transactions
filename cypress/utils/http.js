export const mockError = (url, method) => {
  cy.intercept({
    url,
    method,
  }).as('request');
};

export const mockOk = (url, method, response) => {
  cy.intercept(
    {
      url,
      method,
    },
    {
      fixture: response,
      delay: 50,
    },
  ).as('request');
};

export const mockNoContent = (url, method) => {
  cy.intercept(
    {
      url,
      method,
    },
    {
      body: [],
    },
  ).as('request');
};
