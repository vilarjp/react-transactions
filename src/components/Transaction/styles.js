import styled, { css } from 'styled-components';

export const TransactionWrapper = styled.div`
  ${({ theme }) => css`
    display: flex;
    flex-direction: column;

    padding: ${theme.spacings.medium};
    border-bottom: 1px solid ${theme.colors.lightGray};
    background-color: ${theme.colors.white};
  `}
`;

export const TransactionTitle = styled.div`
  ${({ theme }) => css`
    display: flex;
    justify-content: space-between;
    align-items: center;

    margin-bottom: ${theme.spacings.small};
  `}
`;

export const TransactionUser = styled.p`
  ${({ theme }) => css`
    color: ${theme.colors.darkGray};
    font-weight: ${theme.font.bold};
  `}
`;

export const TransactionStatus = styled.span`
  ${({ theme }) => css`
    color: ${theme.colors.mediumGray};
    font-size: ${theme.font.sizes.small};
  `}
`;

export const TransactionInfo = styled.div`
  display: flex;
  justify-content: space-between;
  align-items: center;
`;

export const TransactionDate = styled.p`
  ${({ theme }) => css`
    color: ${theme.colors.darkGray};
  `}
`;

export const TransactionAmount = styled.span`
  ${({ theme }) => css`
    color: ${theme.colors.mediumBlack};
    font-weight: ${theme.font.bold};
  `}
`;
