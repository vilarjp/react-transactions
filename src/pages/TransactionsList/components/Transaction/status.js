export const statusList = {
  paid: 'Paga',
  refused: 'Recusada',
};
