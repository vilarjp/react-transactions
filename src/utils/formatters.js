export function formatMoney(amount) {
  if (isNaN(amount) || amount === null)
    throw new Error('Invalid amount format');

  return new Intl.NumberFormat('pt-BR', {
    style: 'currency',
    currency: 'BRL',
  }).format(amount);
}

export function formatDate(date) {
  try {
    const dateParsed = new Date(Date.parse(date));
    const checkIfDateIsValid =
      dateParsed.toISOString() === date &&
      dateParsed.toUTCString() === new Date(date).toUTCString();

    if (checkIfDateIsValid) {
      const dateToFormat = new Date(date);
      const formattedDDMMYYYY = new Intl.DateTimeFormat('pt-BR').format(
        dateToFormat,
      );

      const formattedHours = dateToFormat.getHours();
      const formattedMinutes = dateToFormat.getMinutes();

      const timezoneOffSet = dateToFormat.getTimezoneOffset() / 60;
      if (timezoneOffSet === 0)
        return `${formattedDDMMYYYY} ${formattedHours - 3}:${formattedMinutes}`;
      return `${formattedDDMMYYYY} ${formattedHours}:${formattedMinutes}`;
    } else {
      throw new Error('Invalid date format');
    }
  } catch {
    throw new Error('Invalid date format');
  }
}
