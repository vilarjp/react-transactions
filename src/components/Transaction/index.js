import React from 'react';

import { formatMoney, formatDate } from '../../utils/formatters';
import { statusList } from './status';

import * as SC from './styles';

function Transaction({ user, status, date, amount }) {
  return (
    <SC.TransactionWrapper role="transaction ">
      <SC.TransactionTitle>
        <SC.TransactionUser>{user}</SC.TransactionUser>
        <SC.TransactionStatus>{statusList[status] ?? '-'}</SC.TransactionStatus>
      </SC.TransactionTitle>
      <SC.TransactionInfo>
        <SC.TransactionDate>{formatDate(date)}</SC.TransactionDate>
        <SC.TransactionAmount>{formatMoney(amount)}</SC.TransactionAmount>
      </SC.TransactionInfo>
    </SC.TransactionWrapper>
  );
}

export default Transaction;
