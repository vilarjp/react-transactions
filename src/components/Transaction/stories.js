import React from 'react';

import Transaction from '.';

export default {
  title: 'Transaction',
  component: Transaction,
  argTypes: {
    user: {
      type: 'string',
    },
    status: {
      type: 'string',
    },
    date: {
      type: 'date',
    },
    amount: {
      type: 'string',
    },
  },
};

export const Default = (args) => <Transaction {...args} />;

Default.args = {
  user: 'JoÃ£o S Silva',
  status: 'Recusada',
  date: '2020-11-10T19:43:56.451Z',
  amount: '100',
};
