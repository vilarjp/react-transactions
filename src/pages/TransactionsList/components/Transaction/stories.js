import React from 'react';

import Transaction from '.';

export default {
  title: 'Transaction',
  component: Transaction,
  argTypes: {
    user: {
      type: 'string',
    },
    status: {
      control: {
        type: 'inline-radio',
        options: ['paid', 'refused'],
      },
    },
    date: {
      type: 'date',
    },
    amount: {
      type: 'string',
    },
  },
};

export const Default = (args) => <Transaction {...args} />;

Default.args = {
  user: 'JoÃ£o S Silva',
  status: 'refused',
  date: '2020-11-10T19:43:56.451Z',
  amount: '100',
};
