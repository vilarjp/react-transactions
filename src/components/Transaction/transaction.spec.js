import React from 'react';
import { render, screen } from '../../test/render';
import { formatMoney, formatDate } from '../../utils/formatters';
import { statusList } from './status';

import Transaction from '.';

const paidTransaction = {
  user: 'any_user',
  status: 'paid',
  date: '2020-11-10T19:43:56.451Z',
  amount: '123456789',
};

const refusedTransaction = {
  user: 'any_user',
  status: 'refused',
  date: '2020-11-10T19:43:56.451Z',
  amount: '123456789',
};

describe('<Transaction />', () => {
  it('should render the paid transaction with correct values', () => {
    render(<Transaction {...paidTransaction} />);

    expect(screen.getByText(paidTransaction.user)).toBeInTheDocument();
    expect(
      screen.getByText(statusList[paidTransaction.status]),
    ).toBeInTheDocument();
    expect(
      screen.getByText(statusList[paidTransaction.status]),
    ).toMatchSnapshot();
    expect(
      screen.getByText(formatDate(paidTransaction.date)),
    ).toBeInTheDocument();
    expect(
      screen.getByText(formatMoney(paidTransaction.amount)),
    ).toBeInTheDocument();
    expect(
      screen.getByText(formatDate(paidTransaction.date)),
    ).toMatchSnapshot();
    expect(
      screen.getByText(formatMoney(paidTransaction.amount)),
    ).toMatchSnapshot();
  });

  it('should render the refused transaction with correct values', () => {
    render(<Transaction {...refusedTransaction} />);

    expect(
      screen.getByText(statusList[refusedTransaction.status]),
    ).toBeInTheDocument();
    expect(
      screen.getByText(statusList[refusedTransaction.status]),
    ).toMatchSnapshot();
  });

  it('should render empty status if transaction status is not found', () => {
    render(<Transaction {...refusedTransaction} status={'any_wrong_status'} />);

    expect(screen.getByText('-')).toBeInTheDocument();
  });
});
