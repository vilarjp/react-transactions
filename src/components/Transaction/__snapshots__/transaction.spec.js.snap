// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[
  `<Transaction /> should render the paid transaction with correct values 1`
] = `
.c0 {
  color: #72737A;
  font-size: 1.4rem;
}

<span
  class="c0"
>
  Paga
</span>
`;

exports[
  `<Transaction /> should render the paid transaction with correct values 2`
] = `
.c0 {
  color: #454550;
}

<p
  class="c0"
>
  11/10/2020 16:43
</p>
`;

exports[
  `<Transaction /> should render the paid transaction with correct values 3`
] = `
.c0 {
  color: #2B2B2B;
  font-weight: 700;
}

<span
  class="c0"
>
  R$123,456,789.00
</span>
`;

exports[
  `<Transaction /> should render the refused transaction with correct values 1`
] = `
.c0 {
  color: #72737A;
  font-size: 1.4rem;
}

<span
  class="c0"
>
  Recusada
</span>
`;
